deepchem/utils/grover.py

```diff
@@ -0,0 +1,143 @@
+import numpy as np
+from typing import List, Any
+from deepchem.feat.graph_data import BatchGraphData
+
+try:
+    import torch
+except ModuleNotFoundError:
+    pass
+
+
+def _get_atom_scopes(graph_index):
+    """Atom scope is a list of tuples with a single entry for every
+    molecule in the batched graph. The entry indicates the beginning
+    node index for a molecule and the number of nodes in the molecule.
+
+    Example
+    -------
+    >>> import numpy as np
+    >>> graph_index = np.array([0, 0, 1, 1, 1])
+    >>> _get_atom_scopes(graph_index)
+    [[0, 2], [2, 3]]
+    """
+    # graph_index indicates which atom belongs to which molecule
+    mols = np.unique(graph_index)
+    scopes = []
+    for mol in mols:
+        positions = np.where(graph_index == mol, 1, 0)
+        scopes.append([np.argmax(positions), np.count_nonzero(positions)])
+    return scopes
+
+
+def _get_bond_scopes(edge_index, graph_index):
+    """Bond scope is a list of tuples with a single entry for every molecule
+    in the batched graph. The entry indicates the beginning bond index for a
+    molecule and the number of bonds in the molecule.
+
+    Example
+    -------
+    >>> edge_index = np.array([[0, 1, 2, 4], [1, 0, 4, 2]]) # a molecule with 4 bonds
+    >>> graph_index = np.array([0, 0, 1, 1, 1])
+    >>> _get_bond_scopes(edge_index, graph_index)
+    [[0, 2], [2, 2]]
+    """
+    mols = np.unique(graph_index)
+    bond_index = graph_index[edge_index[0]]
+    scopes = []
+    for mol in mols:
+        positions = np.where(bond_index == mol, 1, 0)
+        scopes.append([np.argmax(positions), np.count_nonzero(positions)])
+    return scopes
+
+
+def _compute_b2revb(edge_index):
+    """Every edge in a grover graph is a directed edge. Hence, a bond
+    is represented by two edges of opposite directions. b2revb is a representation
+    which stores for every edge, the index of reverse edge of that edge.
+
+    Example
+    -------
+    >>> import numpy as np
+    >>> edge_index = np.array([[0, 1, 2, 4], [1, 0, 4, 2]])
+    >>> _compute_b2revb(edge_index)
+    [1, 0, 3, 2]
+    """
+    b2revb = [0] * len(edge_index[0])
+    for i, bond in enumerate(edge_index.T):
+        for j, (sa, da) in enumerate(edge_index.T):
+            if sa == bond[1] and da == bond[0]:
+                b2revb[i] = j
+    return b2revb
+
+
+def _get_a2b(n_atoms, edge_index):
+    """a2b is a mapping between atoms and their incoming bonds
```

Comment:
Can you add a `Parameters` section here documenting the types of the argument?

Reply:
added